src/Pipeline.cpp

```diff
@@ -0,0 +1,1057 @@
+#include <algorithm>
+
+#include "Pipeline.h"
+#include "Argument.h"
+#include "Func.h"
+#include "IRVisitor.h"
+#include "LLVM_Headers.h"
+#include "LLVM_Output.h"
+#include "Lower.h"
+#include "Output.h"
+#include "PrintLoopNest.h"
+
+using namespace Halide::Internal;
+
+namespace Halide {
+
+using std::vector;
+using std::string;
+using std::set;
+
+/** An inferred argument. Inferred args are either Params,
+ * ImageParams, or Buffers. The first two are handled by the param
+ * field, and global images are tracked via the buf field. These
+ * are used directly when jitting, or used for validation when
+ * compiling with an explicit argument list. */
+struct InferredArgument {
+    Argument arg;
+    Parameter param;
+    Buffer buffer;
+
+    bool operator<(const InferredArgument &other) const {
+        if (arg.is_buffer() && !other.arg.is_buffer()) {
+            return true;
+        } else if (other.arg.is_buffer() && !arg.is_buffer()) {
+            return false;
+        } else {
+            return arg.name < other.arg.name;
+        }
+    }
+};
+
+struct PipelineContents {
```

Comment:
Is it possible to make this an opaque struct, and move the definition/usages to Pipeline.cpp?